format/logstash.js

```diff
@@ -0,0 +1,42 @@
+const { fromJS } = require('immutable')
+
+const formats = {}
+
+// Source:
+// https://github.com/logstash-plugins/logstash-patterns-core/blob/master/patterns/httpd
+
+formats['HTTPD_COMBINEDLOG'] = (source) => {
+  let headers = {}
+  if (source.agent !== '-') {
+    headers['user-agent'] = source.agent
+  }
+  if (source.referrer !== '-') {
+    headers['referer'] = source.referrer
+  }
+
+  const request = {
+    time: source['@timestamp'],
+    address: source.clientip,
+    method: source.verb,
+    url: source.request,
+    captured_headers: ['user-agent', 'referer'],
+    headers
+  }
+
+  const response = {
+    status: source.response
+  }
+
+  return fromJS({request, response})
+}
+
+function parser (format = 'HTTPD_COMBINEDLOG') {
```

Comment:
The convention so far has been to pass options in a JavaScript object so you can extend it easily later if needed.

Reply:
I'm also not a big fan of coupling the `formats` dictionary with the parser function. The parser should take in input whatever is in the `formats` object. The caller do the `formats` lookup. The same way we do it for nginx formats.

Reply:
@benfle true

Reply:
About the coupling, it's just that the format is directly a transform function. I don't see how it could be done differently.

Reply:
Maybe in that context we should drop the parser and just have:

```
const elasticsearchInput = input.elasticsearch.create({
   parser: format.logstash.formats['HTTPD_COMBINEDLOG']
})
```

Straightforward.

Reply:
Yeah, that's what I meant. Pass the format directly instead of the key (`HTTPD_COMBINEDLOG`)

Reply:
Oh I see what you mean now :) I'm fine with passing the format like you did initially then.

Maybe later we can parse the format the same way we do it with Nginx.